File: Commander/Viewer.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

using Commander.Enums;
using Commander.Extension;

namespace Commander
{
    class Viewer
    {
        public Viewer(Form parent) => this.parent = parent;

        public void SetViewerRatio(double ratio)
        {
            Debugger.Log(1, "Main", $"Viewer Ratio: {ratio}\r\n");
            parent.Invoke((Action)(() =>
            {
                viewer.Height = (int)(parent.ClientSize.Height * ratio);
                viewer.Location = new Point(0, parent.ClientSize.Height - viewer.Height - statusHeight);
            }));
        }

        public void SetStatusRatio(double ratio) => parent.Invoke((Action)(() => statusHeight = (int)(parent.ClientSize.Height * ratio)));

        public void Show(bool show)
        {
            if (show)
            {
                viewer = new Control
                {
                    BackColor = Color.BlueViolet,
                    Size = new Size(parent.ClientSize.Width, 0),
                    Location = new Point(0, parent.ClientSize.Height),
                    Parent = parent,
                    Anchor = (System.Windows.Forms.AnchorStyles)
                                System.Windows.Forms.AnchorStyles.Left
                                | System.Windows.Forms.AnchorStyles.Right
                                | System.Windows.Forms.AnchorStyles.Bottom
                };
                parent.Controls.Add(viewer);
                parent.Controls.SetChildIndex(viewer, 0);
            }
            else
            {
                parent.Controls.Remove(viewer);
                viewer.Dispose();
            }
        }

        readonly Form parent;
        int statusHeight;
        Control viewer;
    }
}

File: Commander/Viewer.fs
namespace Commander

open System.Windows.Forms
open System.Drawing
open System

type Viewer(parent: Control) =
    let mutable viewer: Control option = None
    let mutable statusHeight = 0

    member this.SetViewerRatio(ratio: double) = 
        let setRation () = 
            match viewer with 
            | Some viewer ->
                let size = parent.ClientSize
                viewer.Height <- int (double size.Height * ratio)
                viewer.Location <- new Point(0, size.Height - viewer.Height - statusHeight)
            | None -> ()
        parent.Invoke(Action(setRation)) 

    member this.SetStatusRatio(ratio: double) = 
        let size = parent.ClientSize
        parent.Invoke(Action(fun () -> statusHeight <- int (double size.Height * ratio)))

    member this.SetFile(file: string) =
        let setFile (viewer: Control) = 
            if file = null && viewer.Visible then
                viewer.Visible <- false
            elif file <> null && not viewer.Visible then
                viewer.Visible <- true
            if file <> null && file.EndsWith(".pdf", StringComparison.InvariantCultureIgnoreCase) then
                //let dateTime = DateTime.Now
                //(parent as MainForm).Browser.LostFocus += FocusControl;

                //void FocusControl(object s, EventArgs e)
                //{
                //    if (DateTime.Now - dateTime > TimeSpan.FromSeconds(3))
                //        (parent as MainForm).Browser.LostFocus -= FocusControl;
                //    parent.BeginInvoke((Action)(() => (parent as MainForm).Browser.Focus()));
                //}
                       
                //viewer.Controls.Clear();
                //var browser = new ChromiumWebBrowser("");
                //viewer.Controls.Add(browser);
                ////browser.Load($"file:///{file}");
                //browser.Load($"https://www.caseris.de");
                ()

        match viewer with
        | Some viewer -> parent.Invoke(Action(fun () -> setFile viewer)) |> ignore
        | None -> ()

    member this.Show show =
        match show with
        | true -> 
            let size = parent.ClientSize
            viewer <- Some (new Control(Size = new Size(size.Width, 0),
                                Location = new Point(0, size.Height),
                                Parent = parent,
                                Anchor = (AnchorStyles.Left ||| AnchorStyles.Right ||| AnchorStyles.Bottom)))
            match viewer with 
            | Some viewer ->
                parent.Controls.Add(viewer)
                parent.Controls.SetChildIndex(viewer, 0)
                viewer.Visible <- false
            | None -> ()
        | false ->
            match viewer with 
            | Some viewer ->
                parent.Controls.Remove(viewer)
                viewer.Dispose()
            | None -> ()


